### ISSUE TITLE
Random freezing issue

### ISSUE THREAD

**Issue by jankanb** [pnext/three-loader#46]
I have set up an example implementation of three-loader to display a Potree point cloud. Even for point clouds of less than a million points, after a random moment of navigating around, the browser tab freezes completely. No errors are displayed in the console. When looking at the activity in chrome, the memory is below 500MB. In Chrome when I then force close the GPU process (in Window -> Task Manager), the tab returns to normal but of course WebGL context is gone.

When having larger point clouds the issue seems to appear faster. In all cases the point cloud budget is set to 2 million. Same issues occur when reducing the budget to 1 million.

Would you have any clue about the cause of this issue? Could it be related to disposing of old nodes?

I have tried the latest version of three-loader, and a few older versions as well

From Chrome debug logs I get the following error when the freeze happens:
`GPU hang occurred, msgtracer returned -1`

**Comment by hccampos** [comment]
I have had this issue in the past but haven't been able to pinpoint the cause. Probably disposal of old nodes indeed. I did fix an issue related to that which came from the original potree and which triggered an infinite loop.

**Comment by jankanb** [comment]
It's quite annoying since it happens every time, after a few seconds (1-30) of navigating so basically it's not usable. Is the fix part of the latest version already? Did you do any other changes to avoid the issue from happening? 

On the official potree examples I don't seem have this issue, and their point clouds are sometimes even bigger.

**Comment by jankanb** [comment]
Could the fixes in this PR resolve the issue ? https://github.com/potree/potree/pull/635

**Comment by hccampos** [comment]
Yes, the fixes are in the latest version. I haven't seen this happen that often in our point clouds, even with several million points. I'll try to investigate when I have a moment. But PRs most definitely welcome.

**Comment by hccampos** [comment]
@jankanb I just released v0.1.4 which should fix the random crash. It turns out that when disposing points in the LRU, only the children of the node were being disposed and not the node itself which could lead to an infinite loop.

I also added some of the memory fixes done in the Potree PR you mentioned.

Hopefully this makes it better.

**Comment by jankanb** [comment]
Awesome! I will test the new version tomorrow. I was also testing a point cloud with classification so maybe this was causing the issue to occur faster, despite the lower amount of points.

**Comment by jankanb** [comment]
I just tried the new version but still have the same issue. After some navigation, a freeze... I noticed you didn't add all the memory fixes from Potree, maybe this still has an impact?

**Comment by hccampos** [comment]
I didn't delete the material because otherwise things will break, but it shouldn't be related.

The crash you are experiencing must be something else, probably not even related to memory. Did you check if the CPU usage goes to 100%? Or is it just crashing the page?

I can't reproduce this with any of our massive projects at Pix4D so it might also be some issue related to your code or with the classification which we don't use at all.

**Comment by jankanb** [comment]
The CPU usage of the tab is not high; According to the debug logs it is the GPU process that crashes, which is around 500mb in memory and 10-20% CPU.

I have tried the example included in this repo with one of my point clouds and I have the same freezing issue (so doubt it would be the code). Maybe something is wrong with the point cloud itself? There are two things which could be different:

- There is a Classification attribute in the Potree file, but the points have no classification; So I would think it has no impact
- The Potree is exported in EPSG:3857; maybe the coordinates are too big and creating memory leaks faster? Even when shifting the point cloud to the origin after loading, the freezing occurs.

**Comment by hccampos** [comment]
Oh this leads us into a very dark rabbit hole of GPUs and drivers and configurations 😭 I have a similar thing, sometimes with way more than 500Mb in the GPU process and I have seen no crashes of that kind. But I am on one of the latest MacBook Pro 15" laptops.

This could also be related to ThreeJS. The latest Potree uses a custom renderer and not ThreeJS. Could you try forcing your ThreeJS version down to some other version, for example?

**Comment by jankanb** [comment]
Just tried down to v0.99 (from 107) but still freezing... 😞 

The last variable is the point cloud itself still; Maybe exporting in local coordinate system would help?

**Comment by jankanb** [comment]
I tried the example https://codesandbox.io/s/yw2p3446j9?autoresize=1&view=preview and I also get freezes sometimes there, after a few seconds. (Harder to reproduce consistently, just need to wait a bit). So probably it's not the point cloud/size of coordinates either. 

It seems like the freeze issue does not occur in Safari. I have also tried on two other computers in Chrome and it freezes as well, so I can't be a computer issue.

I did a quick try of the pix4d online point cloud viewer and it seems not to freeze in Chrome at first sight. Is it using the same version of three-loader? Just wondering what could be different then.

**Comment by hccampos** [comment]
Unfortunately I still couldn't reproduce this issue. On our viewers we are running the latest version of both ThreeJS and pnext/three-loader and not really doing anything special with either of them. Maybe there are different settings for initializing the ThreeJS renderer or something.

But this looking like some issue in your combination of Chrome/OS/Drivers/GPU.

**Comment by jankanb** [comment]
For me what is still remarkable is that the issue occurs on different computers (all mac though, but not the same models), but does not occur in the pix4d version in chrome.. So there must be somewhere still something different. 

Also remarkable is that the included example, with default three settings, also freezes. There's not so many settings on the renderer, have tried a few (eg, powerPreference, antialias, ..)

It frustrates me this is so hard to debug..

**Comment by jankanb** [comment]
I have tried the latest develop branch of Potree, and tested one of my point clouds in an example. I don't encounter any freezing in Chrome, and the example is using Three.JS as renderer.

I still think the issue is related to three-loader. The only area where I could find GPU related code is in potree.js updateVisibility function. I already tried reducing the MAX_LOADS_TO_GPU and MAX_NUM_NODES_LOADING to 1 but it has no effect. 

Still strange the issue is not occurring for me on the pix4d version, if it would be the same three-loader version. 

Btw the Chrome version I am using is 76.0.3809.100
